### PR TITLE
Python Timezones (java only) 

### LINKED ISSUE #19: Unpickling datetimes with timezones is not supported
At the moment Pyrolite doesn't support unpickling `datetime.date` or `datetime.datetime` objects  if they have a `tzinfo` property describing their timezone set. I would like to talk about adding support for this feature, but it's a bit tricky because of the different ways you can give a datetime a timezone.

For a datetime like this guy: `datetime.datetime(2014, 7, 8, 10)`, the `pickletools.dis` output looks like this:

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\x00\x00\x00\x00\x00'
   65: p        PUT        1
   68: t        TUPLE      (MARK at 22)
   69: p    PUT        2
   72: R    REDUCE
   73: p    PUT        3
   76: .    STOP
highest protocol among opcodes = 0
```

but for one like this:

```python
from dateutil.tz import tzutc
obj = datetime.datetime(2014, 7, 8, 10, 10, 0, 0, tzutc())
pickletools.dis(obj)
```

the pickle stream looks like this:

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\n\x00\x00\x00\x00'
   63: p        PUT        1
   66: c        GLOBAL     'copy_reg _reconstructor'
   91: p        PUT        2
   94: (        MARK
   95: c            GLOBAL     'dateutil.tz tzutc'
  114: p            PUT        3
  117: c            GLOBAL     'datetime tzinfo'
  134: p            PUT        4
  137: g            GET        4
  140: (            MARK
  141: t                TUPLE      (MARK at 140)
  142: R            REDUCE
  143: p            PUT        5
  146: t            TUPLE      (MARK at 94)
  147: p        PUT        6
  150: R        REDUCE
  151: p        PUT        7
  154: t        TUPLE      (MARK at 22)
  155: p    PUT        8
  158: R    REDUCE
  159: p    PUT        9
  162: .    STOP
```

which has that reconstructor call and then the tuple for the tzinfo instance which Pyrolite doesn't expect at all and fails to deserialize. Worse yet are tzinfos from pytz:

```python
from pytz import timezone
amsterdam = timezone('Europe/Amsterdam')
pickletools.dis(datetime.datetime(2014, 7, 8, 10, 10, 0, 0, amsterdam))
```

gives a pickle stream of 

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\n\x00\x00\x00\x00'
   63: p        PUT        1
   66: c        GLOBAL     'pytz _p'
   75: p        PUT        2
   78: (        MARK
   79: S            STRING     'Europe/Amsterdam'
   99: p            PUT        3
  102: I            INT        1200
  108: I            INT        0
  111: S            STRING     'LMT'
  118: p            PUT        4
  121: t            TUPLE      (MARK at 78)
  122: p        PUT        5
  125: R        REDUCE
  126: p        PUT        6
  129: t        TUPLE      (MARK at 22)
  130: p    PUT        7
  133: R    REDUCE
  134: p    PUT        8
  137: .    STOP
highest protocol among opcodes = 0
```

which has the `pytz._p` object in there with a string representing the zoneinfo name and the offset as well. You can also (and sadly I've seen code in the wild do this) use pytz's UTC representation like so:

```python
import pytz
utc = pytz.utc
pickletools.dis(datetime.datetime(2014, 7, 8, 10, 10, 0, 0, utc))
```

```
    0: c    GLOBAL     'datetime datetime'
   19: p    PUT        0
   22: (    MARK
   23: S        STRING     '\x07\xde\x07\x08\n\n\x00\x00\x00\x00'
   63: p        PUT        1
   66: c        GLOBAL     'pytz _UTC'
   77: p        PUT        2
   80: (        MARK
   81: t            TUPLE      (MARK at 80)
   82: R        REDUCE
   83: p        PUT        3
   86: t        TUPLE      (MARK at 22)
   87: p    PUT        4
   90: R    REDUCE
   91: p    PUT        5
   94: .    STOP
highest protocol among opcodes = 0
```

which is another somewhat annoying variation. Finally, I believe one could in a very twisted world give custom, extended objects as the `tzinfo` property that aren't of the base `dateutil.tz.tzinfo` class or `pytz._p` class. 

My suggestion/plan is to implement support for the second tuple element, and to hardcode an understanding of those two classes which almost everyone uses into the datetime unpickler, and still explode if someone passes a custom class that isn't registered in the java unpickler. If this approach sounds reasonable, I have some further questions:

 - Is there a better way to unpickle that second tuple element than the current approach where we bit shift out members of the byte string? Can I call out to a different piece of the system to get an object I can then inspect to build a java-side Timezone to add to the `Date` object?
 - In the first example above I see escaped newlines coming out in the pickle stream and I don't see that in any of the tests, any idea why my system might be doing that? 
 - Has anyone requested this functionality before or tried this kind of thing?

Thanks for any pointers or help or feedback you can give me!

I discovered this while debugging Apache Spark, so for posterity here's the Spark JIRA: https://issues.apache.org/jira/browse/SPARK-6411

### PR DESCRIPTION
**Pull request by airhorns**
This adds support for depickling most of the commonly used timezone representations from Python in Java land. 

 - Adds a TimeZoneConstructor which converts pickled timezone representations of timezones into java.util.TimeZone objects
 - Adds support for the two argument tuple in the DateTimeConstructor, the inner date’s state and the timezone
 - Add support for depickling python datetime objects with `dateutil.tz.tzutc` objects as their timezone

This is tricky for a couple reasons. For protocol 0 and 1, Python didn’t use `__setstate__` to unpickle objects implemented in C, it used functions from the copy_reg module. So, unlike pytz timezone objects which are implemented in python, the datetime.tzinfo standard library timezone class gets pickled using a “constructor” of `copy_reg._reconstructor` in protocol 0. Further complicating things is that the object that almost everyone uses, `dateutil.tz.tzutc`, is a subclass of this C class. This is pickled as a call to the `_reconstructor` function with three arguments: the subclass or “real” class we are trying to create, the base class which is implemented in C land, and the state that the eventual object we are trying to depickle should have. The reconstructor then allocates a new instance of the base class with `__new__`, but passes the sub class as the class to mark the object as. So the underlying C class does the allocation, but out comes an allocated instance of the fancy sub class. Yeeesh. Then, `_reconstructor` calls the base classes’ `__init__` with the state to actually set the object back up again.

I opted to not really try to implement this logic cause IMO it’s crazy, and a deprecated protocol. Instead, I short cutted. I implemented a Pyrolite land version of the `_reconstructor` function which checks the “sub class” (given to it as a configured IObjectConstructor of some sort) to see if it knows how to reconstruct an object given the “base class” and “state”, which are from the unpickle stack. It uses reflection to check for a `reconstruct` function on the constructor. I then added an implementation of reconstruct to the TimeZoneConstructor so it can check for the very specific conditions that indicate we’re trying to unpickle a `dateutil.tz.tzutc` object, and just leverage the fact that we know it’s UTC and return a UTC TimeZone there. This is all, which means there is no support for unpickling other objects that are subclasses of C classes, like other things in dateutil, like `dateutil.tz.tzlocal`. This is something we could add in the future but requires making the code even more complex and confusing. I think it can be avoided for now since I am the first person to need this functionality, and I know I don’t need tzlocal support, since I feel as though most people stay away from those objects since they don’t support DST properly, and stick with pytz objects, which were easier to implement in the previous commit.

For protocol 2, `__setstate__` is a thing, so I added a Tzinfo object which knows how to be depickled and have it’s state set. The DateTimeConstructor then knows how to be passed a Tzinfo object and ask it for its TimeZone so it can set that on the returned Calendar. Thing is, a `dateutil.tz.tzutc` object doesn’t actually have any state, womp womp. The timezone for it is simply implied by the name of the constructor as far as I can tell. This means that some object somewhere has to know that, so I did something somewhat dirty and make the tzinfo constructor understand when it needs to force itself to UTC. This means that the `__setstate__` is called on the tzinfo object, but with no state, and has no effect, because the object previously learned what java-land TimeZone object it should represent when it was constructed, as that’s the only point in time where we knew the name of the constructor.

Yuck.

@irmen feel free to have a look whenever it's convienient and please let me know if I can change things around, I have not done much java at all. 

@udnay maybe you could give me a quick code review as well if you have time. 

Fixes #19

### CODE REVIEW

**Comment by irmen** [comment]
Wow, I knew timezones are hard but never guessed it would manifest itself like this.
Will take a look later, and see how this may or may not be translated to the .net part of pyrolite.
Thanks for your effort so far!

**Comment by irmen** [comment]
We'll need support on the Pickler side as well I guess put_calendar should choose one of the two timezone representations for Python. Or perhaps add a config item to choose one or the other (in Pickler and in Config)

File: java/src/net/razorvine/pickle/objects/TimeZoneConstructor.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+
+	private Object createInfoFromDateutilTzutc(Object[] args) {
+		// In the case of the dateutil.tz.tzutc constructor, which is a python subclass of the datetime.tzinfo class, there is no state
+		// to set, because the zone is implied by the constructor. Pass the timezone indicated by hte constructor here
+		return new Tzinfo(TimeZone.getTimeZone("UTC"));
+	}
+
+	private Object createZoneFromPytz(Object[] args) {
+
+		if (args.length != 4 && args.length != 1)
```
Comment (udnay) [inline java/src/net/razorvine/pickle/objects/TimeZoneConstructor.java]: Not sure which is wrong, this or the exception text below.

File: java/src/net/razorvine/pickle/objects/Tzinfo.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Timezone offset class that implements __setstate__ for the unpickler
+ * to track what TimeZone a dateutil.tz.tzoffset or tzutc should unpickle to
+ */
+public class Tzinfo {
+
+    private boolean forceTimeZone;
+    private TimeZone timeZone;
+
+    public Tzinfo(TimeZone forceTimeZone) {
```
Comment (udnay) [inline java/src/net/razorvine/pickle/objects/Tzinfo.java]: Can you rename this to just `timeZone`, the name swap below is strange.

File: java/src/net/razorvine/pickle/objects/TimeZoneConstructor.java (hunk 1 above)
Comment (irmen) [inline java/src/net/razorvine/pickle/objects/TimeZoneConstructor.java]: Reading on, the code and the comment below are talking about either 4 or 1 arg. So it seems that the exception text has a typo.

File: java/src/net/razorvine/pickle/objects/Tzinfo.java (hunk 2 above)
Comment (irmen) [inline java/src/net/razorvine/pickle/objects/Tzinfo.java]: I agree

**Comment by airhorns** [comment]
Ok, added support for pickling `java.util.Calendar`s with `TimeZone`s attached. Since the standard library for Python doesn’t have a concrete tzinfo subclass, we’re forced to do something kind of bad and pickle Java TimeZone instances as calls to the pytz.tzinfo unpickler function. This means if you want to unpickle a datetime with a timezone from Pyrolite, you need to use pytz. I think at the end of the day this is actually ok: pytz is linked to rather prominently from the Python stdlib documentation as the go to tool for managing timezones in Python, and anecdotally I haven’t seen any open source projects which don’t use pytz for the tzinfo implementation there. I am totally open to other ideas here.

**Comment by irmen** [comment]
I guess sometime in the future we'll have to add support for java8's java.time api as well...

**Pull request merged**